### PR TITLE
Fix issue #686 RemoveRow slice bounds out of range

### PR DESCRIPTION
**Pull request by liserjrqlxue**
# PR Details

修复`RemoveRow`删除最后一行的报错
## Description

`adjustHyperlinks`通过
```xlsx.Hyperlinks.Hyperlink = append(xlsx.Hyperlinks.Hyperlink[:rowIdx],xlsx.Hyperlinks.Hyperlink[rowIdx+1:]...)```
来跳过`rowIdx`对应行  
但是当`rowIdx>=len(xlsx.Hyperlinks.Hyperlink)`时`xlsx.Hyperlinks.Hyperlink[rowIdx+1:]`下标超限
所以在进行如`RemoveRow`删除最后一行等相关操作时会报如`slice bounds out of range [94:93]`错误
## Related Issue

https://github.com/360EntSecGroup-Skylar/excelize/issues/686
## Motivation and Context

修复`RemoveRow`最后一行时`slice bounds out of range`错误
## How Has This Been Tested

我用[patch-1](https://github.com/liserjrqlxue/excelize/tree/patch-1)替代excelize，修复了对最后一行删除的错误
https://github.com/liserjrqlxue/filterExcel/commit/748a89d8600f2a0d5f1b08279d276fe87309b8b0

捕获

捕获

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: adjust.go [hunk 1]
```diff
@@ -115,8 +115,10 @@ func (f *File) adjustHyperlinks(xlsx *xlsxWorksheet, sheet string, dir adjustDir
 			if (dir == rows && num == rowNum) || (dir == columns && num == colNum) {
 				f.deleteSheetRelationships(sheet, linkData.RID)
 				if len(xlsx.Hyperlinks.Hyperlink) > 1 {
-					xlsx.Hyperlinks.Hyperlink = append(xlsx.Hyperlinks.Hyperlink[:rowIdx],
-						xlsx.Hyperlinks.Hyperlink[rowIdx+1:]...)
+					if rowIdx<len(xlsx.Hyperlinks.Hyperlink){
```
Comment (xuri) [inline adjust.go:118]: `rowId` should be always less than the length of the `xlsx.Hyperlinks.Hyperlink`, that is confusing, could you provide a spreadsheet xlsx attachment without confidential info? I need to investigate more details about this issue.

Comment (liserjrqlxue) [reply adjust.go:118]: I can't upload file with error message:'Something went really wrong, and we can;t process thatfile. Try again.'
But I wrote a code to build test xlsx and reproduce this issue:
```
package main

import (
	"fmt"

	"github.com/360EntSecGroup-Skylar/excelize/v2"
	"github.com/liserjrqlxue/goUtil/simpleUtil"
)

func main() {
	var excel =excelize.NewFile()
	var sheet1= excel.GetSheetName(0)
	simpleUtil.CheckErr(excel.SetCellValue(sheet1,"A1",""))
	simpleUtil.CheckErr(excel.SetCellValue(sheet1,"A2",""))
	simpleUtil.CheckErr(excel.SetCellValue(sheet1,"B1",""))
	simpleUtil.CheckErr(excel.SetCellValue(sheet1,"B2",""))
	simpleUtil.CheckErr(excel.SetCellHyperLink(sheet1,"A1","","External"))
	simpleUtil.CheckErr(excel.SetCellHyperLink(sheet1,"A2","","External"))
	simpleUtil.CheckErr(excel.SetCellHyperLink(sheet1,"B2","","External"))
	simpleUtil.CheckErr(excel.SaveAs("debug.xlsx"))
	var rows,err= excel.GetRows(sheet1)
	simpleUtil.CheckErr(err)
	fmt.Printf("len(rows)=%d\n",len(rows))
	simpleUtil.CheckErr(excel.RemoveRow(sheet1,len(rows)))
}
```
After build and run, result may be like below:
捕获

I also wrote a test code which you can run in this repo use `go test -run TestRemoveRowDebug`:
```
func TestRemoveRowDebug(t *testing.T){
	f := NewFile()
	sheet1 := f.GetSheetName(0)
	assert.NoError(t, f.SetCellValue(sheet1,"A1",""))
	assert.NoError(t, f.SetCellValue(sheet1,"A2",""))
	assert.NoError(t, f.SetCellValue(sheet1,"B1",""))
	assert.NoError(t, f.SetCellValue(sheet1,"B2",""))
	assert.NoError(t, f.SetCellHyperLink(sheet1,"A1","","External"))
	assert.NoError(t, f.SetCellHyperLink(sheet1,"A2","","External"))
	assert.NoError(t, f.SetCellHyperLink(sheet1,"B2","","External"))
	assert.NoError(t, f.RemoveRow(sheet1,2))
}
```
The result may be like below:
捕获

Comment (xuri) [reply adjust.go:118]: Thanks for your feedback, I propose to fix this issue like this:

```go
// order is important
if offset < 0 {
    for i := 0; i < len(xlsx.Hyperlinks.Hyperlink); i++ {
        linkData := xlsx.Hyperlinks.Hyperlink[i]
        colNum, rowNum, _ := CellNameToCoordinates(linkData.Ref)

        if (dir == rows && num == rowNum) || (dir == columns && num == colNum) {
            f.deleteSheetRelationships(sheet, linkData.RID)
            if len(xlsx.Hyperlinks.Hyperlink) > 1 {
                xlsx.Hyperlinks.Hyperlink = append(xlsx.Hyperlinks.Hyperlink[:i],
                    xlsx.Hyperlinks.Hyperlink[i+1:]...)
                i--
            } else {
                xlsx.Hyperlinks = nil
            }
        }
    }
}
```

Comment (liserjrqlxue) [reply adjust.go:118]: In your code, it seems my fix only avoid error but not handle hyperlinks correctly.  
I think use `i` decreasing as 'order is important' may also fix this issue:
```golang
// order is important
if offset < 0 {
    for i :=  len(xlsx.Hyperlinks.Hyperlink)-1; i < 0; i-- {
        linkData := xlsx.Hyperlinks.Hyperlink[i]
        colNum, rowNum, _ := CellNameToCoordinates(linkData.Ref)

        if (dir == rows && num == rowNum) || (dir == columns && num == colNum) {
            f.deleteSheetRelationships(sheet, linkData.RID)
            if len(xlsx.Hyperlinks.Hyperlink) > 1 {
                xlsx.Hyperlinks.Hyperlink = append(xlsx.Hyperlinks.Hyperlink[:i],
                    xlsx.Hyperlinks.Hyperlink[i+1:]...)
            } else {
                xlsx.Hyperlinks = nil
            }
        }
    }
}
```

**Pull request merged**